File: 1.6.js
// Given an image represented by a NxN matrix, where
// each pixel in the image is 4 bytes, write a method
// to rotate the image by 90 degrees. Can you do this in place?

const m = [
  [1, 2, 3, 4],
  [5, 6, 7, 8],
  [9, 10, 11, 12], 
  [13, 14, 15, 16]
]

const rotateMatrix = (m) => {
  let min = 0;
  let max = m.length-1;
  while (min < max && max >= m.length/2) {
    let dif = 0;
    while (dif+min < max) {
      const temp1 = m[min][min+dif];
      const temp2 = m[min+dif][max];
      const temp3 = m[max][max-dif];
      const temp4 = m[max-dif][min];

      m[min][min+dif] = temp2;
      m[min+dif][max] = temp3;
      m[max][max-dif] = temp4;
      m[max-dif][min] = temp1;
      
      dif++;
    }
    min++;
    max--;
  }
  return m;
};
console.log(rotateMatrix(m));

File: 1.6.rb
# Given an image represented by a NxN matrix, where
# each pixel in the image is 4 bytes, write a method
# to rotate the image by 90 degrees. Can you do this in place?

def rotate matrix 
  min = 0;
  max = matrix.length-1

  while min < max and max >= matrix.length/2
    dif = 0
    while min + dif < max
      t1 = matrix[min][min+dif]
      t2 = matrix[min+dif][max]
      t3 = matrix[max][max-dif]
      t4 = matrix[max-dif][min]
       
      matrix[min][min+dif] = t4
      matrix[min+dif][max] = t1
      matrix[max][max-dif] = t2
      matrix[max-dif][min] = t3

      dif += 1
    end
    min += 1
    max -= 1
  end
  matrix
end

m = [
  [1, 2, 3, 4],
  [5, 6, 7, 8],
  [9, 10, 11, 12], 
  [13, 14, 15, 16]
]

p rotate(m)
